crypto/asn1/i2d_pr.c

```diff
@@ -30,35 +30,19 @@ int i2d_PrivateKey(const EVP_PKEY *a, unsigned char **pp)
         }
         return ret;
     }
-    if (a->keymgmt != NULL) {
+    if (evp_pkey_is_provided(a)) {
         /* The private key includes everything */
+        OSSL_ENCODER_CTX *ctx;
         int selection =
             OSSL_KEYMGMT_SELECT_ALL_PARAMETERS | OSSL_KEYMGMT_SELECT_KEYPAIR;
-        OSSL_ENCODER_CTX *ctx =
-            OSSL_ENCODER_CTX_new_by_EVP_PKEY(a, "DER", selection, NULL, NULL);
-        BIO *out = BIO_new(BIO_s_mem());
-        BUF_MEM *buf = NULL;
+        size_t length;
         int ret = -1;
 
-        if (ctx != NULL
-            && out != NULL
+        if ((ctx = OSSL_ENCODER_CTX_new_by_EVP_PKEY(a, "DER", selection,
+                                                    NULL, NULL)) != NULL
             && OSSL_ENCODER_CTX_get_num_encoders(ctx) != 0
-            && OSSL_ENCODER_to_bio(ctx, out)
-            && BIO_get_mem_ptr(out, &buf) > 0) {
-            ret = buf->length;
-
-            if (pp != NULL) {
-                if (*pp == NULL) {
-                    *pp = (unsigned char *)buf->data;
-                    buf->length = 0;
-                    buf->data = NULL;
-                } else {
-                    memcpy(*pp, buf->data, ret);
-                    *pp += ret;
-                }
-            }
-        }
-        BIO_free(out);
+            && OSSL_ENCODER_to_data(ctx, pp, &length))
```

Comment:
Hmm, thinking about this - there is a problem with the length value passed in when the *pp is non-NULL. The old API was simply of the strcpy() unbounded write kind so the right thing would be to set the length to some "really high" value. But perhaps some reasonable limit should be used?

Reply:
Thanks for the reminder.  I'm setting it to something really large, rather than a limit that we might regret in a few years.